File: py/sort/linked_list/SelectionSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class SelectionSort:
    def sort(self, head: ListNode) -> None:
        prev = head
        while prev.next:
            current, node = prev, prev.next
            while node.next:
                if current.next.value > node.next.value:
                    current = node
                node = node.next

            if prev != current:
                next = current.next
                current.next, next.next, prev.next = next.next, prev.next, next
            prev = prev.next

File: java/sort/linked_list/SelectionSort.java
package sort.linked_list;

public class SelectionSort {
    public void sort(ListNode head) {
        ListNode prev = head;
        while (null != prev.next) {
            ListNode current = prev;
            for (ListNode node = prev.next; null != node.next; node = node.next) {
                if (current.next.value > node.next.value) {
                    current = node;
                }
            }

            if (prev != current) {
                ListNode next = current.next;
                current.next = next.next;
                next.next = prev.next;
                prev.next = next;
            }
            prev = prev.next;
        }
    }

    public class ListNode {
        public int value = 0;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}
